File: backend/app/schemas/permission.py
from pydantic import BaseModel, Field


class PermissionEntry(BaseModel):
    department_id: int
    # Display convenience only -- always ignored on write, always
    # populated by permission_service.get_matrix on read.
    department_code: str | None = None
    page_key: str
    access_level: str = Field(pattern="^(none|read|write)$")


class PermissionMatrixUpdate(BaseModel):
    entries: list[PermissionEntry]


class PermissionPage(BaseModel):
    key: str
    label: str

File: frontend/src/types/permission.ts
export type AccessLevel = 'none' | 'read' | 'write'

export interface PermissionEntry {
  department_id: number
  /** Display convenience, always populated on read -- see
   * types/department.ts for the authoritative Department master shape. */
  department_code?: string
  page_key: string
  access_level: AccessLevel
}

/** The calling user's own effective access per page -- page_key -> level. */
export type MyPermissions = Record<string, AccessLevel>

/** A governable page's key + display label, as returned by
 * /api/permissions/pages -- the backend is the only source of this
 * list, there is no hardcoded frontend copy. */
export interface PermissionPage {
  key: string
  label: string
}
